Common/src/main/groovy/io/github/groovymc/cgl/api/datagen/recipe/GRecipeProvider.groovy

```diff
@@ -0,0 +1,257 @@
+        recipe(new GCookingRecipeBuilder(RecipeSerializer.BLASTING_RECIPE), clos)
+    }
+
+    protected GCookingRecipeBuilder campfireCooking(@DelegatesTo(value = GCookingRecipeBuilder, strategy = Closure.DELEGATE_FIRST) @ClosureParams(value = SimpleType, options = 'io.github.groovymc.cgl.api.datagen.GCookingRecipeBuilder') Closure clos) {
+        recipe(new GCookingRecipeBuilder(RecipeSerializer.CAMPFIRE_COOKING_RECIPE), clos)
+    }
+
+    protected GCookingRecipeBuilder smoking(@DelegatesTo(value = GCookingRecipeBuilder, strategy = Closure.DELEGATE_FIRST) @ClosureParams(value = SimpleType, options = 'io.github.groovymc.cgl.api.datagen.GCookingRecipeBuilder') Closure clos) {
+        recipe(new GCookingRecipeBuilder(RecipeSerializer.SMOKING_RECIPE), clos)
+    }
+
+    protected GShapedRecipeBuilder shaped(@DelegatesTo(value = GShapedRecipeBuilder, strategy = Closure.DELEGATE_FIRST) @ClosureParams(value = SimpleType, options = 'io.github.groovymc.cgl.api.datagen.GShapedRecipeBuilder') Closure clos) {
+        recipe(new GShapedRecipeBuilder(), clos)
+    }
+
+    protected GShapelessRecipeBuilder shapeless(@DelegatesTo(value = GShapelessRecipeBuilder,strategy = Closure.DELEGATE_FIRST) @ClosureParams(value = SimpleType, options = 'io.github.groovymc.cgl.api.datagen.GShapelessRecipeBuilder') Closure clos) {
+        recipe(new GShapelessRecipeBuilder(), clos)
+    }
+
+    protected <T extends SaveableRecipe> T recipe(@DelegatesTo.Target('recipe') Class<T> recipeClass, @DelegatesTo(target = 'recipe', genericTypeIndex = 0, strategy = Closure.DELEGATE_FIRST) @ClosureParams(FirstParam.FirstGenericType) Closure clos) {
+        return recipe(recipeClass.getDeclaredConstructor().newInstance(), clos)
+    }
+
+    protected <T extends SaveableRecipe> T recipe(@DelegatesTo.Target('recipe') T recipe, @DelegatesTo(target = 'recipe', strategy = Closure.DELEGATE_FIRST) @ClosureParams(FirstParam) Closure clos) {
+        recipe.provider = this
+        clos.resolveStrategy = Closure.DELEGATE_FIRST
+        clos.delegate = recipe
+        clos(recipe)
+        forgottenRecipes.add(recipe)
+        return recipe
+    }
+
+    protected <T extends RecipeBuilder> SaveableRecipe normal(@DelegatesTo.Target('recipe') T recipe, @DelegatesTo(target = 'recipe', strategy = Closure.DELEGATE_FIRST) @ClosureParams(FirstParam) Closure clos = {}) {
+        clos.resolveStrategy = Closure.DELEGATE_FIRST
+        clos.delegate = recipe
+        clos(recipe)
+        return new SaveableRecipe() {
+            @Override
+            void save(ResourceLocation location) {
+                forgottenRecipes.remove(this)
+                recipe.save(getProvider().writer, location)
+            }
+
+            @Override
+            GRecipeProvider getProvider() {
+                return GRecipeProvider.this
+            }
+
+            @Override
+            void setProvider(GRecipeProvider provider) {
+
+            }
+        }
+    }
+}
+
+@CompileStatic
```

Comment:
Hmm, can traits not have their own class files? I'm really not liking this bit with more than one top level class in a class file - makes finding references to stuff in GitHub a bit difficult